packages/oruga/src/composables/useInputHandler.ts

```diff
@@ -224,13 +229,81 @@ export function useInputHandler<T extends ValidatableFormElement>(
         emits("invalid", event);
     }
 
+    /**
+     * Provides a way to force the watcher on `updateCustomValidationMessage` to re-run
+     *
+     * There are some cases (e.g. changes to the element's validation attributes) that can
+     * force changes to the element's `validityState`, which isn't a reactive property.
+     * We call `updateCustomValidationMessage` in cases where we expect relevant changes
+     * to `validityState`. That's enough to update the DOM correctly. However, changes in
+     * `validityState` are likely to cause the function version of `props.customValidity`
+     * to execute different branches that may have different dependencies on reactive
+     * properties. The `watchEffect` call on `updateCustomValidationMessage` needs to know
+     * about those new reactive dependencies, and the only way to inform it of them is to
+     * re-run `updateCustomValidationMessage` within the watcher's context.
+     *
+     * This does mean that if you need to update the DOM element's validity state _immediately_,
+     * you'll need to call updateCustomValidationMessage directly (to make the change happen now
+     * rather than waiting for the reactivity system) _and_ trigger this `ref` (to register
+     * reactive dependencies), so `updateCustomValidationMessage` will run twice for those updates.
+     */
+    const forceValidationReactivityUpdate = ref(null);
+
+    /**
+     * Propagates any custom constraint validation message to the underlying DOM element
+     */
+    function updateCustomValidationMessage(): void {
+        forceValidationReactivityUpdate.value;
+        if (!(props.useHtml5Validation ?? true)) {
+            return;
+        }
+        const element = maybeElement.value;
+        if (!isDefined(element)) {
+            return;
+        }
+        const validity = props.customValidity ?? "";
+        if (typeof validity === "string") {
+            element.setCustomValidity(validity);
+        } else {
+            // The custom validation message may depend on `element.validity`,
+            // which isn't a reactive property. `element.validity` depends on
+            // the element's current value and the native constraint validation
+            // attributes. We can use `props.modelValue` as a reasonable proxy
+            // for the DOM element's value, and `props.modelValue` _is_ reactive,
+            // so we can use it to help solve that reactivity problem.
+            // When `props.modelValue` is retrieved in this function, even if we
+            // throw the value away, the `watchEffect` call on this function will
+            // register it as a dependency.
+            props.modelValue;
+            element.setCustomValidity(validity(element.validity));
```

Comment:
If you call `props.modelValue` here in any way, just pass it as the second parameter to the `validity` function. That way the user can use it, or just use the first parameter without the second.

Reply:
I discovered that getting the types to line up was slightly less painful than I thought, although the strength of the actual type checking is pretty limited. I've pushed a fix for this.